### PR TITLE
Replaced factories with new templated factory

### LINKED ISSUE #1143: Replace existing factories with new templated factory
### Description of the task

Once #1120 is complete, we will have a generic factory class that we can use everywhere, without having to make a new factory every time we need one. Now we can replace our existing factories with this new general one.

Existing factories that should be replaced
* BackendFactory
* PlayFactory

### Acceptance criteria

- [ ] Replace the BackendFactory with the new generic factory
- [ ] Replace the BackendFactory with the new generic factory

### Blocked By

#1120

### PR DESCRIPTION
**Pull request by amrm3lm**
### Description

replaced uses of play_factory and backend_factory with the templated generic_factory. 
removed play_factory and backend_factory files. 
renamed generic_factory.cpp to generic_factory.tpp
### Testing Done

everything builds and tests pass
### Resolved Issues

resolves #1143 
### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by kelvinkoon — COMMENTED** [review]
The formatting test seems to not be passing. Try and run the ./fix_formatting.sh script and it should pass that.

File: src/software/ai/hl/stp/play/corner_kick_play.cpp [hunk 1]
```diff
@@ -235,5 +235,5 @@ void CornerKickPlay::updatePassGenerator(PassGenerator &pass_generator)
     pass_generator.setPasserPoint(world.ball().position());
 }
 
-// Register this play in the PlayFactory
-static TPlayFactory<CornerKickPlay> factory;
+// Register this play in the genericFactory
+static TGenericFactory<std::string, Play, CornerKickPlay> factory;
```
Comment (MathewMacDougall) [inline src/software/ai/hl/stp/play/corner_kick_play.cpp]: nit: missing newline at end of file

File: src/software/ai/hl/stp/play/enemy_freekick_play.cpp [hunk 2]
```diff
@@ -132,5 +132,5 @@ void EnemyFreekickPlay::getNextTactics(TacticCoroutine::push_type &yield)
     } while (true);
 }
 
-// Register this play in the PlayFactory
-static TPlayFactory<EnemyFreekickPlay> factory;
+// Register this play in the genericFactory
+static TGenericFactory<std::string, Play, EnemyFreekickPlay> factory;
```
Comment (MathewMacDougall) [inline src/software/ai/hl/stp/play/enemy_freekick_play.cpp]: nit: missing newline at end of file

File: src/software/ai/hl/stp/play/example_play.cpp [hunk 3]
```diff
@@ -66,5 +66,5 @@ void ExamplePlay::getNextTactics(TacticCoroutine::push_type &yield)
     } while (true);
 }
 
-// Register this play in the PlayFactory
-static TPlayFactory<ExamplePlay> factory;
+// Register this play in the genericFactory
+static TGenericFactory<std::string, Play, ExamplePlay> factory;
```
Comment (MathewMacDougall) [inline src/software/ai/hl/stp/play/example_play.cpp]: nit: missing newline at end of file

File: src/software/ai/hl/stp/play/free_kick_play.cpp [hunk 4]
```diff
@@ -283,5 +283,5 @@ void FreeKickPlay::shootOrFindPassStage(
              shoot_tactic->hasShotAvailable());
 }
 
-// Register this play in the PlayFactory
-static TPlayFactory<FreeKickPlay> factory;
+// Register this play in the genericFactory
+static TGenericFactory<std::string, Play, FreeKickPlay> factory;
```
Comment (MathewMacDougall) [inline src/software/ai/hl/stp/play/free_kick_play.cpp]: nit: missing newline at end of file

File: src/software/ai/hl/stp/play/halt_play.cpp [hunk 5]
```diff
@@ -38,5 +38,5 @@ void HaltPlay::getNextTactics(TacticCoroutine::push_type &yield)
     } while (true);
 }
 
-// Register this play in the PlayFactory
-static TPlayFactory<HaltPlay> factory;
+// Register this play in the genericFactory
+static TGenericFactory<std::string, Play, HaltPlay> factory;
```
Comment (MathewMacDougall) [inline src/software/ai/hl/stp/play/halt_play.cpp]: nit: missing newline at end of file

File: src/software/ai/hl/stp/play/kickoff_enemy_play.cpp [hunk 6]
```diff
@@ -152,5 +153,5 @@ void KickoffEnemyPlay::getNextTactics(TacticCoroutine::push_type &yield)
     } while (true);
 }
 
-// Register this play in the PlayFactory
-static TPlayFactory<KickoffEnemyPlay> factory;
+// Register this play in the genericFactory
+static TGenericFactory<std::string, Play, KickoffEnemyPlay> factory;
```
Comment (MathewMacDougall) [inline src/software/ai/hl/stp/play/kickoff_enemy_play.cpp]: nit: missing newline at end of file

File: src/software/ai/hl/stp/play/kickoff_friendly_play.cpp [hunk 7]
```diff
@@ -145,5 +146,6 @@ void KickoffFriendlyPlay::getNextTactics(TacticCoroutine::push_type &yield)
     }
 }
 
-// Register this play in the PlayFactory
-static TPlayFactory<KickoffFriendlyPlay> factory;
+
+// Register this play in the genericFactory
+static TGenericFactory<std::string, Play, KickoffFriendlyPlay> factory;
```
Comment (MathewMacDougall) [inline src/software/ai/hl/stp/play/kickoff_friendly_play.cpp]: nit: missing newline at end of file

File: src/software/ai/hl/stp/play/penalty_kick_enemy_play.cpp [hunk 8]
```diff
@@ -61,5 +62,5 @@ void PenaltyKickEnemyPlay::getNextTactics(TacticCoroutine::push_type &yield)
     } while (true);
 }
 
-// Register this play in the PlayFactory
-static TPlayFactory<PenaltyKickEnemyPlay> factory;
+// Register this play in the genericFactory
+static TGenericFactory<std::string, Play, PenaltyKickEnemyPlay> factory;
```
Comment (MathewMacDougall) [inline src/software/ai/hl/stp/play/penalty_kick_enemy_play.cpp]: nit: missing newline at end of file

File: src/software/ai/hl/stp/play/penalty_kick_play.cpp [hunk 9]
```diff
@@ -91,5 +92,5 @@ void PenaltyKickPlay::getNextTactics(TacticCoroutine::push_type &yield)
     } while (true);
 }
 
-// Register this play in the PlayFactory
-static TPlayFactory<PenaltyKickPlay> factory;
+// Register this play in the genericFactory
+static TGenericFactory<std::string, Play, PenaltyKickPlay> factory;
```
Comment (MathewMacDougall) [inline src/software/ai/hl/stp/play/penalty_kick_play.cpp]: nit: missing newline at end of file

File: src/software/ai/hl/stp/play/shoot_or_chip_play.cpp [hunk 10]
```diff
@@ -165,4 +166,5 @@ void ShootOrChipPlay::getNextTactics(TacticCoroutine::push_type &yield)
     } while (!shoot_or_chip_tactic->done());
 }
 
-static TPlayFactory<ShootOrChipPlay> factory;
+// Register this play in the genericFactory
+static TGenericFactory<std::string, Play, ShootOrChipPlay> factory;
```
Comment (MathewMacDougall) [inline src/software/ai/hl/stp/play/shoot_or_chip_play.cpp]: nit: missing newline at end of file

File: src/software/ai/hl/stp/play/stop_play.cpp [hunk 11]
```diff
@@ -133,5 +135,5 @@ void StopPlay::getNextTactics(TacticCoroutine::push_type &yield)
     } while (true);
 }
 
-// Register this play in the PlayFactory
-static TPlayFactory<StopPlay> factory;
+// Register this play in the genericFactory
+static TGenericFactory<std::string, Play, StopPlay> factory;
```
Comment (MathewMacDougall) [inline src/software/ai/hl/stp/play/stop_play.cpp]: nit: missing newline at end of file

File: src/software/ai/hl/stp/play/test_plays/halt_test_play.cpp [hunk 12]
```diff
@@ -35,5 +35,5 @@ void HaltTestPlay::getNextTactics(TacticCoroutine::push_type &yield)
     } while (true);
 }
 
-// Register this play in the PlayFactory
-static TPlayFactory<HaltTestPlay> factory;
+// Register this play in the genericFactory
+static TGenericFactory<std::string, Play, HaltTestPlay> factory;
```
Comment (MathewMacDougall) [inline src/software/ai/hl/stp/play/test_plays/halt_test_play.cpp]: nit: missing newline at end of file

File: src/software/ai/hl/stp/play/test_plays/move_test_play.cpp [hunk 13]
```diff
@@ -37,5 +37,5 @@ void MoveTestPlay::getNextTactics(TacticCoroutine::push_type &yield)
     } while (!move_test_tactic_center_field->done());
 }
 
-// Register this play in the PlayFactory
-static TPlayFactory<MoveTestPlay> factory;
+// Register this play in the genericFactory
+static TGenericFactory<std::string, Play, MoveTestPlay> factory;
```
Comment (MathewMacDougall) [inline src/software/ai/hl/stp/play/test_plays/move_test_play.cpp]: nit: missing newline at end of file

File: src/software/backend/grsim_backend.cpp [hunk 14]
```diff
@@ -57,5 +57,5 @@ void GrSimBackend::updateGrSim()
     }
 }
 
-// Register this backend in the BackendFactory
-static TBackendFactory<GrSimBackend> factory;
+// Register this play in the genericFactory
+static TGenericFactory<std::string, Backend, GrSimBackend> factory;
```
Comment (MathewMacDougall) [inline src/software/backend/grsim_backend.cpp]: nit: missing newline at end of file

File: src/software/backend/radio_backend.cpp [hunk 15]
```diff
@@ -32,5 +32,5 @@ void RadioBackend::receiveWorld(World world)
     Subject<World>::sendValueToObservers(world);
 }
 
-// Register this backend in the BackendFactory
-static TBackendFactory<RadioBackend> factory;
+// Register this play in the genericFactory
+static TGenericFactory<std::string, Backend, RadioBackend> factory;
```
Comment (MathewMacDougall) [inline src/software/backend/radio_backend.cpp]: nit: missing newline at end of file

File: src/software/backend/simulator_backend.cpp [hunk 16]
```diff
@@ -139,5 +139,5 @@ void SimulatorBackend::runSimulationLoop(World world)
     }
 }
 
-// Register this backend in the BackendFactory
-static TBackendFactory<SimulatorBackend> factory;
+// Register this play in the genericFactory
+static TGenericFactory<std::string, Backend, SimulatorBackend> factory;
```
Comment (MathewMacDougall) [inline src/software/backend/simulator_backend.cpp]: nit: missing newline at end of file

File: src/software/util/design_patterns/generic_factory_test.cpp [hunk 17]
```diff
@@ -1,4 +1,5 @@
-#include "software/util/design_patterns/generic_factory.cpp"
+
```
Comment (MathewMacDougall) [inline src/software/util/design_patterns/generic_factory_test.cpp]: nit: extra line at top of file

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
basically all nit comments, changes themselves look good :+1:

**Review by garethellis0 — COMMENTED** [review]
Nothing to add beyond Mat's comments.

**Comment by amrm3lm** [comment]
@MathewMacDougall I've added new lines locally and pushed them, however they dont seem to be reflected in githubs file viewer, but then again no files seem to have a new line at the end when viewed on github?

**Comment by MathewMacDougall** [comment]
Yea, it's not super obvious but the missing newlines have been pushed. 

GitHub flags missing newlines with a little red arrow in the diff like:
image

And all I'm doing is looking for that marker when I leave the comments. When you pushed your changes to add the newlines, the markers went away.

It's mostly a style thing, but here's why it's generally good to always have newlines at the end of a file (and why GitHub "complains" about it): https://stackoverflow.com/questions/5813311/no-newline-at-end-of-file

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1:

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Review by kelvinkoon — APPROVED** [review]
LGTM!

**Pull request merged**